File: euler/benchmarks/problem12/julia/main.jl
# run with julia main.jl
# julia version used: 0.6.3
module BenchmarkProblem12

mutable struct PrimeGenerator
    n::Int64
    lastn::Int64
    sieve::Dict{Int64,Int64}
    PrimeGenerator() = new(3, 2, Dict())
end

function nextPrime(self::PrimeGenerator)::Int64

    while haskey(self.sieve, self.n)
        prime = self.sieve[self.n]

        delete!(self.sieve, self.n)

        composite = self.n + prime + prime
        while haskey(self.sieve, composite)
            composite += prime + prime
        end
        self.sieve[composite] = prime

        self.n += 2
    end

    self.sieve[self.n * self.n] = self.n
    r = self.lastn
    self.lastn = self.n
    self.n += 2
    r
end

struct FactorFinder
    known::Dict{Int64,Set{Int64}}
    nextPrimes::PrimeGenerator
    knownPrimes::Vector{Int64}
    FactorFinder() = new(Dict(1 => Set(1)), PrimeGenerator(), [])
end

function getPrimeFactors(self::FactorFinder, of::Int64)::Vector{Int64}
    factors = Int64[]
    quotient = of

    for prime in self.knownPrimes
        if prime > quotient
            return factors
        end

        remainder = rem(quotient, prime)
        while remainder == 0
            quotient = div(quotient, prime)
            remainder = rem(quotient, prime)
            push!(factors, prime)
        end
    end

    while true
        prime = nextPrime(self.nextPrimes)
        push!(self.knownPrimes, prime)

        if prime > quotient
            return factors
        end

        remainder = rem(quotient, prime)
        while remainder == 0
            quotient = div(quotient, prime)
            remainder = rem(quotient, prime)
            push!(factors, prime)
        end
    end

    factors
end

function getFactors(self::FactorFinder, of::Int64)::Set{Int64}
    if haskey(self.known, of)
        return self.known[of]
    end

    factors = Set{Int64}([1, of])
    for prime in getPrimeFactors(self, of)
        factor = div(of, prime)
        for subfactor in getFactors(self, factor)
            push!(factors, subfactor)
        end
    end

    self.known[of] = factors
    factors
end

function solve(factorCount)::Int64
    finder = FactorFinder()
    adder = 0
    tn = 0

    while true
        adder += 1
        tn += adder

        factors = getFactors(finder, tn)
        if length(factors) > factorCount
            return tn
        end
    end
end

function main()
    println("Answer: " * string(solve(1000)))
    return 0
end

end

File: euler/benchmarks/problem12/javascript/main.js

class PrimeGenerator
{
    constructor() {
        this.n = 3;
        this.last = 2;
        this.sieve = new Map();
    }

    next() {
        let prime = 0;
        while (this.sieve.has(this.n))
        {
            prime = this.sieve.get(this.n);

            this.sieve.delete(this.n);

            let composite = this.n + prime + prime;
            while (this.sieve.has(composite))
                composite += prime + prime;
            this.sieve.set(composite, prime);

            this.n = this.n + 2;
        }

        this.sieve.set(this.n * this.n, this.n);
        const r = this.last;
        this.last = this.n;
        this.n += 2;
        return r;
    }
}

class FactorFinder
{
    constructor() {
        this.known = new Map();
        this.nextPrimes = new PrimeGenerator();
        this.knownPrimes = [];
        const init = new Set();
        init.add(1);
        this.known.set(1, init);
    }

    getPrimeFactors(of)
    {
        const factors = [];
        let quotient = of;

        for (let prime of this.knownPrimes)
        {
            if (prime > quotient)
                return factors;

            let remainder = quotient % prime;
            while (remainder == 0)
            {
                quotient /= prime;
                remainder = quotient % prime;
                factors.push(prime);
            }
        }

        for (;;)
        {
            const prime = this.nextPrimes.next();
            this.knownPrimes.push(prime);

            if (prime > quotient)
                return factors;

            let remainder = quotient % prime;
            while (remainder == 0)
            {
                quotient /= prime;
                remainder = quotient % prime;
                factors.push(prime);
            }
        }
    }

    getFactors(of)
    {
        if (this.known.has(of))
            return this.known.get(of);

        const factors = new Set();
        factors.add(1);
        factors.add(of);

        for (const prime of this.getPrimeFactors(of))
        {
            const factor = of / prime;
            for (const subfactor of this.getFactors(factor))
            {
                factors.add(subfactor);
            }
        }

        this.known.set(of, factors);
        return factors;
    }
}

function solve()
{
    const finder = new FactorFinder();
    let adder = 0;
    let tn = 0;

    for (;;)
    {
        adder += 1;
        tn += adder;
        const factors = finder.getFactors(tn);
        if (factors.size > 1000)
            return tn;
    }
}

const answer = solve();
console.log(`Answer: ${answer}`);